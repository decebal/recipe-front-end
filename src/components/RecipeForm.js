import React from 'react';
import PropTypes from 'prop-types';
import { withFormik } from 'formik';
import * as Yup from 'yup';
import {
  Button, ButtonGroup, H2, Hr, Row,
} from '@bootstrap-styled/v4/dist/@bootstrap-styled/v4';
import { lensPath, range, view } from 'ramda';
import TextInput from './form/TextInput';
import Textarea from './form/Textarea';

const formikEnhancer = withFormik({
  validationSchema: Yup.object().shape({
    name: Yup.string()
      .min(2, 'The recipe name must be longer than 2 characters')
      .required('Recipe Name is required.'),
    description: Yup.string()
      .min(2, 'The recipe description must be longer than 2 characters')
      .required('Description is required.'),
    ingredients: Yup.array().of(Yup.string()
      .min(2, 'The recipe ingredient must be longer than 2 characters'))
      .min(2, 'The recipe must have at least 2 ingredients')
      .required('Ingredients are required!'),
    instructions: Yup.array().of(Yup.string()
      .min(2, 'The recipe instruction step must be longer than 2 characters'))
      .min(2, 'The recipe must have at least 2 steps')
      .required('Ingredients are required!'),
  }),

  mapPropsToValues: ({ recipe }) => ({
    ...recipe,
  }),

  handleSubmit: (payload, { setSubmitting, props }) => {
    const { onSubmit } = props;
    setSubmitting(true);
    onSubmit(payload);
    setSubmitting(false);
  },
  displayName: 'RecipeEditor',
});

const RecipeForm = ({
  values,
  touched,
  errors,
  dirty,
  handleChange,
  handleBlur,
  handleSubmit,
  handleReset,
  isSubmitting,
}) => (
  <form onSubmit={handleSubmit}>
    <TextInput
      id="name"
      type="text"
      label="Recipe Name"
      placeholder="Recipe Name"
      error={touched.name && errors.name}
      value={values.name}
      onChange={handleChange}
      onBlur={handleBlur}
    />
    <Textarea
      id="description"
      type="text"
      label="Recipe Description"
      placeholder="Describe the recipe in a few words"
      error={touched.description && errors.description}
      value={values.description}
      onChange={handleChange}
      onBlur={handleBlur}
    />
    <Hr />
    <Row>
      <H2>Ingredients</H2>
    </Row>
    {range(0, 9).map(key => (
      <TextInput
        key={`input-${key}`}
        id={`ingredients[${key}]`}
        type="text"
        placeholder="Ingredient"
        label={`Ingredient no. ${key + 1}`}
        error={view(lensPath(['ingredients', key]), touched) && view(lensPath(['ingredients', key]), errors)}
        value={view(lensPath(['ingredients', key]), values)}
        onChange={handleChange}
        onBlur={handleBlur}
      />
    ))}
    <Hr />
    <Row>
      <H2>Instructions</H2>
    </Row>
    {range(0, 9).map(key => (
      <TextInput
        key={`input-${key}`}
        id={`instructions[${key}]`}
        type="text"
        placeholder="Instruction"
        label={`Step no. ${key + 1}`}
        error={view(lensPath(['instructions', key]), touched) && view(lensPath(['instructions', key]), errors)}
        value={view(lensPath(['instructions', key]), values)}
        onChange={handleChange}
        onBlur={handleBlur}
      />
    ))}
    <Hr />
    <Row style={{ flexFlow: 'row-reverse' }}>
      <ButtonGroup>
        <Button color="secondary" onClick={handleReset} disabled={!dirty || isSubmitting}>Reset</Button>
        <Button color="primary" disabled={isSubmitting} type="submit" id="submit">Submit</Button>
      </ButtonGroup>
    </Row>
    <Hr />
  </form>
);


RecipeForm.propTypes = {
  values: PropTypes.shape({
    name: PropTypes.string,
    description: PropTypes.string,
    ingredients: PropTypes.arrayOf(PropTypes.string),
    instructions: PropTypes.arrayOf(PropTypes.string),
  }),
  touched: PropTypes.shape({
    name: PropTypes.bool,
    description: PropTypes.bool,
    ingredients: PropTypes.arrayOf(PropTypes.bool),
    instructions: PropTypes.arrayOf(PropTypes.bool),
  }),
  errors: PropTypes.shape({
    name: PropTypes.string,
    description: PropTypes.string,
    ingredients: PropTypes.arrayOf(PropTypes.string),
    instructions: PropTypes.arrayOf(PropTypes.string),
  }),
  dirty: PropTypes.bool,
  isSubmitting: PropTypes.bool,
  handleChange: PropTypes.func.isRequired,
  handleBlur: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  handleReset: PropTypes.func.isRequired,
};

RecipeForm.defaultProps = {
  values: {
    name: '',
    description: '',
    ingredients: [],
    instructions: [],
  },
  touched: {
    name: false,
    description: false,
    ingredients: [],
    instructions: [],
  },
  errors: {
    name: null,
    description: null,
    ingredients: [],
    instructions: [],
  },
  dirty: false,
  isSubmitting: false,
};


export default formikEnhancer(RecipeForm);
