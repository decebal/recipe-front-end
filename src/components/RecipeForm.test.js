import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import RecipeForm from './RecipeForm';


describe('<RecipeForm />', () => {
  // it('Test submit form event', () => {
  //   const mockCallBack = jest.fn();
  //
  //   const button = shallow((<RecipeForm onClick={mockCallBack} />));
  //   button.find({ id: "submit" }).simulate('click');
  //   expect(mockCallBack.mock.calls.length).toEqual(1);
  // });

  test('RecipeForm renders correctly', () => {
    const component = renderer
      .create(
        <RecipeForm
          onSubmit={()=>{}}
        />
      );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
