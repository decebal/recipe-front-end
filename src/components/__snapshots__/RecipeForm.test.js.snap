// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecipeForm /> RecipeForm renders correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="name"
    >
      Recipe Name
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="name"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Recipe Name"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-EHOje jYocTM"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="description"
    >
      Recipe Description
    </label>
    <textarea
      className="sc-ifAKCX miDGs"
      id="description"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Describe the recipe in a few words"
      type="text"
      value={null}
    />
  </div>
  <hr
    className="Hr-sc-1a12p8r-0 jPQocB"
  />
  <div
    className="Row-sc-2hg243-0 gRqPSy row"
  >
    <h2
      className="H2-kplljn-0 bPmkUD"
    >
      Ingredients
    </h2>
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="ingredients[0]"
    >
      Ingredient no. 1
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="ingredients[0]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Ingredient"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="ingredients[1]"
    >
      Ingredient no. 2
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="ingredients[1]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Ingredient"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="ingredients[2]"
    >
      Ingredient no. 3
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="ingredients[2]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Ingredient"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="ingredients[3]"
    >
      Ingredient no. 4
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="ingredients[3]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Ingredient"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="ingredients[4]"
    >
      Ingredient no. 5
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="ingredients[4]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Ingredient"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="ingredients[5]"
    >
      Ingredient no. 6
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="ingredients[5]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Ingredient"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="ingredients[6]"
    >
      Ingredient no. 7
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="ingredients[6]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Ingredient"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="ingredients[7]"
    >
      Ingredient no. 8
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="ingredients[7]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Ingredient"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="ingredients[8]"
    >
      Ingredient no. 9
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="ingredients[8]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Ingredient"
      type="text"
      value=""
    />
  </div>
  <hr
    className="Hr-sc-1a12p8r-0 jPQocB"
  />
  <div
    className="Row-sc-2hg243-0 gRqPSy row"
  >
    <h2
      className="H2-kplljn-0 bPmkUD"
    >
      Instructions
    </h2>
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="instructions[0]"
    >
      Step no. 1
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="instructions[0]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Instruction"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="instructions[1]"
    >
      Step no. 2
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="instructions[1]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Instruction"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="instructions[2]"
    >
      Step no. 3
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="instructions[2]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Instruction"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="instructions[3]"
    >
      Step no. 4
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="instructions[3]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Instruction"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="instructions[4]"
    >
      Step no. 5
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="instructions[4]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Instruction"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="instructions[5]"
    >
      Step no. 6
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="instructions[5]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Instruction"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="instructions[6]"
    >
      Step no. 7
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="instructions[6]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Instruction"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="instructions[7]"
    >
      Step no. 8
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="instructions[7]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Instruction"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bxivhb gCIOfD"
  >
    <label
      className="sc-bdVaJa iXjSzF"
      htmlFor="instructions[8]"
    >
      Step no. 9
    </label>
    <input
      className="sc-htpNat bPsheZ"
      id="instructions[8]"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Instruction"
      type="text"
      value=""
    />
  </div>
  <hr
    className="Hr-sc-1a12p8r-0 jPQocB"
  />
  <div
    className="Row-sc-2hg243-0 gRqPSy row"
    style={
      Object {
        "flexFlow": "row-reverse",
      }
    }
  >
    <div
      className="ButtonGroup-sc-1wk454k-0 euUImV btn-group"
      role="group"
    >
      <button
        className="Button-h8watj-0 OlpfH btn disabled btn-secondary"
        onClick={[Function]}
        type="button"
      >
        Reset
      </button>
      <button
        className="Button-h8watj-0 OlpfH btn btn-primary"
        id="submit"
        onClick={[Function]}
        type="submit"
      >
        Submit
      </button>
    </div>
  </div>
  <hr
    className="Hr-sc-1a12p8r-0 jPQocB"
  />
</form>
`;
